File: api/schemas/matrix.py
"""Schemas served by the causal matrix route."""

from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel

# ---- /matrix/frame -------------------------------------------------------


class MatrixRow(BaseModel):
    """One stable constraint row in a delivery day's bounded SF rectangle."""

    constraint_key: str
    constraint_name: str
    contingency_name: str | None = None
    constraint_type: str | None = None
    forecast_mu: float
    ercot_dam_mu: float | None = None
    daily_rank: int
    binding_hours: int
    max_abs_sf: float


class MatrixColumn(BaseModel):
    """One stable settlement-point column in a bounded SF rectangle."""

    settlement_point: str
    settlement_point_type: str | None = None
    load_zone: str | None = None
    max_abs_sf: float


class MatrixSfValues(BaseModel):
    """Row-major recovered implied-SF values aligned to ``rows`` and ``columns``.

    A cell whose ``|value|`` equals the frame's ``sf_abs_cap`` was pinned there
    by the ridge fit's clip and is a bound rather than a measurement (0145). No
    parallel mask is sent: the clip is exact, so ``abs(v) >= sf_abs_cap`` is the
    same test the server would apply, at a fraction of the payload.
    """

    row_count: int
    column_count: int
    values: list[float]


class MatrixFrame(BaseModel):
    """A causal, immutable-artifact-backed Matrix frame for one delivery hour."""

    available: bool
    unavailable_reason: str | None = None
    run_id: str
    delivery_date: date
    interval_ts: datetime
    fit_window_start: datetime | None = None
    fit_window_end: datetime | None = None
    # Which axis got the "primary list" (ranked + searched + bounded) treatment.
    # The wire shape is unchanged — ``rows`` are always constraints and
    # ``columns`` always settlement points — but the client reads this to decide
    # the visual orientation (``nodes`` renders nodes as rows by transposing).
    orientation: Literal["constraints", "nodes"] = "constraints"
    dam_status: Literal["pending", "partial", "available"] = "pending"
    row_ordering: str = "daily_abs_forecast_contribution_desc_then_constraint_key"
    column_ordering: str = "max_abs_sf_desc_then_settlement_point"
    # The bounded rectangle is intentionally not the whole artifact.  Clients
    # need this distinction before describing any visible-only sum.
    rows_truncated: bool = False
    columns_truncated: bool = False
    # Counts describe the artifact universe, not a congestion total.  A client
    # can therefore say "30 of 143 constraints" without implying that the
    # visible rows account for all nodal congestion.
    total_constraint_count: int = 0
    total_settlement_point_count: int = 0
    # Day-wide extrema keep the color legend stable when a client filters the
    # bounded view. Contribution uses the forecast-day range for both sources
    # so their colors remain directly comparable.
    sf_day_max_abs: float = 0.0
    contribution_day_max_abs: float = 0.0
    # The fit's |SF| clip. Cells sitting exactly here were pinned by the ridge
    # rather than measured, and the client marks them; sent so the threshold has
    # one source (compute.sf_map.model.fit.SF_ABS_CAP) instead of a hardcoded 1.0 on both
    # sides of the wire.
    sf_abs_cap: float = 0.0
    rows: list[MatrixRow] = []
    columns: list[MatrixColumn] = []
    sf: MatrixSfValues

File: web/src/api/types/matrix.ts
// =============================================================================
// /matrix/frame — one bounded, day-stable constraint × settlement-point frame.
// The API owns the row and column ordering.  `sf.values` is row-major and is
// aligned exactly to `rows` then `columns`; it contains recovered implied shift
// factors, not an official ERCOT shift-factor field.
// =============================================================================

export type MatrixDamStatus = "pending" | "partial" | "available";

// Which axis the backend gave the primary ranked/searched list. The wire shape
// is unchanged (`rows` are always constraints, `columns` always settlement
// points); the client transposes `nodes` at draw time.
export type MatrixOrientation = "constraints" | "nodes";

export interface MatrixRow {
  constraint_key: string;
  constraint_name: string;
  contingency_name: string | null;
  constraint_type: string | null;
  forecast_mu: number;
  ercot_dam_mu: number | null;
  daily_rank: number;
  binding_hours: number;
  max_abs_sf: number;
}

export interface MatrixColumn {
  settlement_point: string;
  settlement_point_type: string | null;
  load_zone: string | null;
  max_abs_sf: number;
}

// A cell whose |value| reaches the frame's `sf_abs_cap` was pinned there by the
// ridge fit's clip — a bound, not a measurement. No parallel mask is sent: the
// clip is exact, so `Math.abs(v) >= sf_abs_cap` is the same test the server
// would apply, at a fraction of the payload.
export interface MatrixSfValues {
  row_count: number;
  column_count: number;
  values: number[];
}

export interface MatrixFrame {
  available: boolean;
  unavailable_reason: string | null;
  run_id: string;
  delivery_date: string;
  interval_ts: string;
  fit_window_start: string | null;
  fit_window_end: string | null;
  orientation: MatrixOrientation;
  dam_status: MatrixDamStatus;
  row_ordering: string;
  column_ordering: string;
  rows_truncated: boolean;
  columns_truncated: boolean;
  total_constraint_count: number;
  total_settlement_point_count: number;
  sf_day_max_abs: number;
  contribution_day_max_abs: number;
  // The fit's |SF| clip, so the threshold has one source rather than a
  // hardcoded 1.0 on both sides of the wire.
  sf_abs_cap: number;
  rows: MatrixRow[];
  columns: MatrixColumn[];
  sf: MatrixSfValues;
}

// Compact realized-range wire format. `sp_ids` is the one settlement-point
// index for the response; each entry's same-length arrays align to it.
export interface ErcotRangeEntry {
  interval_ts: string;
  system_lambda: number | null;
  congestion: Array<number | null>;
  spp: Array<number | null>;
}

export interface ErcotRangeResponse {
  start: string;
  end: string;
  count: number;
  sp_ids: string[];
  entries: ErcotRangeEntry[];
}
